File: Python/Regular Expression Matching.py
class Solution:
    # @param s, a string
    # @param p, a string
    # @return a boolean
    def isMatch(self, s, p):
        matrix = [[ False for i in range(len(p)+1)] for j in range(len(s)+1)]
        matrix[0][0] = True
        isValid = False
        for i in range(2, len(p)+1, 2):
            if p[i-1] == "*":
                isValid = True
                matrix[0][i] = True;
            else:
                break
        
        for i in range(1,len(p)+1):
            for j in range(1,len(s)+1):
                if s[j-1] == p[i-1] or p[i-1] == "." :
                    matrix[j][i] = matrix[j-1][i-1]
                elif p[i-1] == "*":
                    if s[j-1] == p[i-2] or p[i-2] == ".":
                        matrix[j][i] = matrix[j-1][i] or matrix[j][i-2]
                    else:
                        matrix[j][i] = matrix[j][i-2]

        return matrix[len(s)][len(p)]

File: Java/src/Regular Expression Matching.java
public class Solution {
    public boolean isMatch(String s, String p) {
        boolean matrix[][] = new boolean[s.length()+1][p.length()+1];
        matrix[0][0] = true;
        boolean valid = false;
        for(int j = 2;j <= p.length();j+=2){
            if(p.charAt(j-1)=='*'){ valid = true; matrix[0][j] = true;}
            else{ break;}
        }
        for(int i = 1;i <= s.length();i++){
            for(int j = 1;j <= p.length();j++){
                if(s.charAt(i-1)==p.charAt(j-1) || p.charAt(j-1)=='.') matrix[i][j] = matrix[i-1][j-1];
                else if(p.charAt(j-1)=='*'){
                    if(s.charAt(i-1)==p.charAt(j-2) || p.charAt(j-2)=='.')
                        matrix[i][j] = matrix[i-1][j] || matrix[i][j-2];
                    else
                        matrix[i][j] = matrix[i][j-2];
                }
            }
        }
        return matrix[s.length()][p.length()];
    }
}